test/dynamo/test_misc.py

```diff
@@ -3765,6 +3765,100 @@ def fn(x, y):
         res = opt_fn(x, y)
         self.assertTrue(same(ref, res))
 
+    @unittest.skipIf(sys.version_info < (3, 11), "requires Python 3.11+")
+    def test_py311_jump_offset(self):
+        new_inst = bytecode_transformation.create_instruction
+        consts = (None, 1, 2, 3, 4)
+
+        def create_test_code(jump_opname, target_idx):
+            targets = [
+                new_inst("LOAD_CONST", 1),
+                new_inst("LOAD_CONST", 3),
+            ]
+            jump_to_target_inst = new_inst(jump_opname, target=targets[target_idx])
+            """
+            pseudocode of generated bytecode:
+            def test_py311_fn():
+                goto target1
+            target0:
+                return 1
+            target1:
+                goto [target0/target2] (via fwd or bwd jump)
+                return 2
+            target2:
+                return 3
+                return 4
+            """
+            insts = [
+                new_inst("RESUME", 0),
+                new_inst("JUMP_FORWARD", target=jump_to_target_inst),
+                targets[0],
+                new_inst("RETURN_VALUE"),
+                jump_to_target_inst,
+                new_inst("LOAD_CONST", 2),
+                new_inst("RETURN_VALUE"),
+                targets[1],
+                new_inst("RETURN_VALUE"),
+                new_inst("LOAD_CONST", 4),
+                new_inst("RETURN_VALUE"),
+            ]
+            code_options = collections.OrderedDict(
+                [
+                    ("co_argcount", 0),
+                    ("co_posonlyargcount", 0),
+                    ("co_kwonlyargcount", 0),
+                    ("co_nlocals", 1),
+                    ("co_stacksize", 1),
+                    ("co_flags", 3),
+                    ("co_code", b""),
+                    ("co_consts", consts),
+                    ("co_names", ("test_local",)),
+                    ("co_varnames", ("test_local",)),
+                    ("co_filename", __file__),
+                    ("co_name", "test_py311_fn"),
+                    ("co_qualname", "test_py311_fn"),
+                    ("co_firstlineno", 1),
+                    ("co_linetable", b""),
+                    ("co_exceptiontable", b""),
+                    ("co_freevars", ()),
+                    ("co_cellvars", ()),
+                ]
+            )
+            return bytecode_transformation.clean_and_assemble_instructions(
+                insts,
+                list(code_options.keys()),
+                code_options,
+            )
+
+        # format: jump_opname, target_idx, expected forward jump, expected return value
+        test_args = (
```

Comment:
From this test, do I understand correctly that JUMP_FORWARD and JUMP_BACKWARD can be used the same way and the instructor creation will transform it in the right JUMP_* depending on the target?

Reply:
Yes.